openmls/src/extensions/mod.rs

```diff
@@ -195,6 +195,183 @@ pub enum Extension {
+        let mut hash_map = HashSet::new();
+
+        for extension_type in candidate.iter().map(Extension::extension_type) {
+            // ... trying to insert every element ...
+            if !hash_map.insert(extension_type) {
+                // ... and returning false if an element of the same extension type was inserted before.
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl TryFrom<Vec<Extension>> for Extensions {
+    type Error = &'static str;
+
+    fn try_from(value: Vec<Extension>) -> Result<Self, Self::Error> {
+        if Extensions::validate(&value) {
+            Ok(Self { inner: value })
+        } else {
+            Err("List of extensions must not contain duplicate extension types.")
+        }
+    }
+}
+
+impl Extensions {
+    /// Get a reference to the inner value.
+    pub fn inner(&self) -> &Vec<Extension> {
+        &self.inner
+    }
+
+    /// Get a reference to the [`ApplicationIdExtension`] if there is any.
+    pub fn application_id(&self) -> Option<&ApplicationIdExtension> {
+        // Safety: `.unwrap()` is safe here.
+        self.inner
+            .iter()
+            .find(|e| e.extension_type() == ExtensionType::ApplicationId)
+            .map(|e| e.as_application_id_extension().unwrap())
+    }
+
+    /// Get a reference to the [`RatchetTreeExtension`] if there is any.
+    pub fn ratchet_tree(&self) -> Option<&RatchetTreeExtension> {
+        // Safety: `.unwrap()` is safe here.
+        self.inner
+            .iter()
+            .find(|e| e.extension_type() == ExtensionType::RatchetTree)
+            .map(|e| e.as_ratchet_tree_extension().unwrap())
+    }
+
+    /// Get a reference to the [`RequiredCapabilitiesExtension`] if there is any.
+    pub fn required_capabilities(&self) -> Option<&RequiredCapabilitiesExtension> {
+        // Safety: `.unwrap()` is safe here.
+        self.inner
+            .iter()
+            .find(|e| e.extension_type() == ExtensionType::RequiredCapabilities)
+            .map(|e| e.as_required_capabilities_extension().unwrap())
+    }
+
+    /// Get a reference to the [`ExternalPubExtension`] if there is any.
+    pub fn external_pub(&self) -> Option<&ExternalPubExtension> {
+        // Safety: `.unwrap()` is safe here.
+        self.inner
+            .iter()
+            .find(|e| e.extension_type() == ExtensionType::ExternalPub)
+            .map(|e| e.as_external_pub_extension().unwrap())
+    }
+
+    /// Get a reference to the [`CapabilitiesExtension`] if there is any.
+    pub fn capabilities(&self) -> Option<&CapabilitiesExtension> {
+        // Safety: `.unwrap()` is safe here.
+        self.inner
+            .iter()
+            .find(|e| e.extension_type() == ExtensionType::Capabilities)
+            .map(|e| e.as_capabilities_extension().unwrap())
+    }
+
+    /// Get a reference to the [`LifetimeExtension`] if there is any.
+    pub fn lifetime(&self) -> Option<&LifetimeExtension> {
```

Comment:
Why not have a generic lookup function where you have a parameter of type `ExtensionType` to look up the type you want? This is a rather long hard-coded list.

Having a generic lookup would also be a first step in the direction of being able to use proprietary extensions later on.

Reply:
Do you mean the generic lookup function as a replacement for the provided methods or as a way to refactor this a bit?

The former, i.e., only having an `ExtensionType -> Extension` function, would require the caller to further call `as_<extension_type>()` and take care of (unnecessary) error handling.